### PR TITLE
feat(appsync): expose s3 location api

### PR DESCRIPTION
**Pull request by BryanPan342**
**[ISSSUE]**
s3 location api for cloudformations is not exposed

**[APPROACH]**
Expose s3 location api through `schemaDefinitionFile` property

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by BryanPan342** [comment]
**[TODO]**
Implement `Schema` class to allow for schema definition as suggested by @asterikx:

`schema: appsync.Schema.fromInline(schema)`

`schema: appsync.Schema.fromAsset(path)`

`schema: appsync.Schema.fromBucket(bucket, key[, objectVersion])`

```ts
class Schema extends Construct {
  
  // No definition for now because of code-first implementation
  fromInline(): Schema {
    return new Schema(SchemaDefinition.CODE, '');
  }

  fromFile(file: string): Schema {
    return new Schema(SchemaDefinition.FILE, readFileSync(file).toString('UTF-8'));
  }

  public readonly mode: SchemaDefinition;

  public readonly schema: CfnGraphQLSchema;

  private constructor(mode: SchemaDefinition, definition: string) {
    this.mode = mode;
    this.schema = new CfnGraphQLSchema(this, 'Schema', {
      apiId: <?>,     // Not sure how we can pass in the apiId here.. if anything we can leave it blank and fill it in later
      definition,
    });
  }
}
```

**Comment by BryanPan342** [comment]
**[ACTIONABLES]**
- [ ] chore up '==' to '==='
- [ ] README changes to put a header on example
- [ ] README change s3 example to use s3 bucket location
- [ ] Explanation on use case

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts [hunk 1]
```diff
@@ -704,11 +711,16 @@ export class GraphQLApi extends Construct {
       definition = '';
     } else if ( this.schemaMode == SchemaDefinition.CODE && file) {
       throw new Error('definition mode CODE is incompatible with file definition. Change mode to FILE/S3 or unconfigure schemaDefinitionFile');
+    } else if ( this.schemaMode == SchemaDefinition.S3 && !file) {
+      throw new Error('schemaDefinitionFile must be configured if using S3 definition mode.');
+    } else if ( this.schemaMode == SchemaDefinition.S3 && file ) {
+      definitionS3Location = file;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:717]:
```suggestion
    } else if ( this.schemaMode === SchemaDefinition.S3 && !file) {
      throw new Error('schemaDefinitionFile must be configured if using S3 definition mode.');
    } else if ( this.schemaMode === SchemaDefinition.S3 && file ) {
      definitionS3Location = file;
```

seems like this needs to be fixed up in this entire conditional block

File: packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+function error {
+  printf "\e[91;5;81m$@\e[0m\n"
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh:4]: curious: what's happening here?

File: packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  echo "#######################################################################"
+}
+
+if [[ -z $1 || -z $2 ]]; then
+  error "Error: verification requires [APIKEY] [ENDPOINT]"
+  usage
+  exit 1
+fi
+
+echo THIS TEST SHOULD SUCCEED
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh:19]: describe the test that is about to run and the output that is expected.

File: packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+curl -XPOST -H "Content-Type:application/graphql" -H "x-api-key:$1" -d '{ "query": "query { getTests { version } }" }" }' $2
+echo ""
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh:21]: perhaps indicate success here too?

File: packages/@aws-cdk/aws-appsync/README.md [hunk 5]
```diff
@@ -159,4 +159,25 @@ api.grantMutation(role, 'updateExample');
 
 // For custom types and granular design
 api.grant(role, appsync.IamResource.ofType('Mutation', 'updateExample'), 'appsync:GraphQL');
-```
\ No newline at end of file
+```
+
+### s3-assets
+
+`aws-appsync` supports using s3 location for schema definition.
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/README.md:166]: clarify: 

does this mean it's supporting an S3 location to be referenced and used?
i.e. user already has an existing schema in S3 and they want to point to it

alternatively, does it mean users can author their schema definition locally and 
the cdk will upload to s3 and instrument it as an asset?

I think they're 2 different use cases, but I did get guidance from @rix0rrr that
we should address the former and ask the question as to whether the latter is 
really needed. Related resource where we had that discussion was for a Step Functions
state machine definition, which also supports an S3 location for defining a state machine.

File: packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh (hunk 2 above)
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh:4]: It helps format the error message. I got it from the buildup/builddown script

Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/verify.integ.graphql-s3.sh:4]: Basically, the way it works is the \e specifies an escape sequence.

So bascially this function has three components:
- `\e[91;5;81m$@` - formatting the input text
- `\e[0m` - remove all formatting attributes (sanity for future texts)
- `\n` - newline

Breaking down the first one `\e[91;5;81m$@`:
- `91` - the color red
- `5`- blinking 
- `81` - honestly I have no idea what this is for.. removing it doesn't change anything for me
- `$@` - input to function

**Pull request closed without merging**